### PR TITLE
VTimezone Recurrence parsing

### PR DESCRIPTION
**Pull request by dbondarchuk**
Added parsing VTimezone props to extract the offset for the specific date

This code is built with the help of https://github.com/mangstadt/biweekly as a reference.

It can be a ground-up work for adding support for the recurring events

### CODE REVIEW

**Comment by dbondarchuk** [comment]
@timheerwagen let's have a look on proper timezones :)

File: package.json [hunk 1]
```diff
@@ -49,6 +52,7 @@
     "access": "public"
   },
   "dependencies": {
+    "@datastructures-js/priority-queue": "^6.3.0",
```
Comment (timheerwagen) [inline package.json:55]: Is there a way without this extra dependency? Why exactly do we need this?

File: src/lib/parse/utils/recurrence/compoundIterator.ts
Comment (timheerwagen) [inline src/lib/parse/utils/recurrence/compoundIterator.ts:1]: Is there a reason why you use classes? I would prefer functions.

File: src/lib/parse/utils/recurrence/dateTimeBuilder.ts
Comment (timheerwagen) [inline src/lib/parse/utils/recurrence/dateTimeBuilder.ts:1]: Same here.

File: src/lib/parse/utils/recurrence/filter.ts
Comment (timheerwagen) [inline src/lib/parse/utils/recurrence/filter.ts:1]: Why are you using bitwise here? Are there any alternatives?

File: src/lib/parse/utils/recurrence/generators.ts [hunk 2]
```diff
@@ -0,0 +1,1078 @@
...
+    filter,
+    yearGenerator,
+    monthGenerator,
+    dayGenerator,
+    hourGenerator,
+    minuteGenerator,
+    secondGenerator,
+  );
+
+  // TODO(msamuel): does this work?
```
Comment (timheerwagen) [inline src/lib/parse/utils/recurrence/generators.ts:819]: Please remove unused comments from the other package

File: src/lib/parse/utils/timezone/offset.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  const minuteStr = matches[4];
+  const minuteOffset = !minuteStr ? 0 : parseInt(minuteStr, 10);
+
+  return { positive, hourOffset, minuteOffset };
+};
+
+export const hoursToMillis = (hours: number): number => hours * 60 * 60 * 1000;
+
+export const minutesToMillis = (minutes: number): number => minutes * 60 * 1000;
```
Comment (timheerwagen) [inline src/lib/parse/utils/timezone/offset.ts:27]: Just use hoursToMilliseconds and minutesToMilliseconds from date-fns for consistency please.

**Review by timheerwagen — CHANGES_REQUESTED** [review]
A huge thanks for this comprehensive Contribution. Due to the fact that recurrence has now been introduced, it would be of advantage to use the whole thing for calendars, events etc. as well.

The utils for generating the recurrence and timezone would certainly be better off in src/utils, instead of src/lib/parse/utils. Because I would also like to provide a function which can be used independently similar to "https://github.com/jkbrzt/rrule" rrule.all or rrule.between.

The function could have following type: (rule: VEventRecurrenceRule, options: {start?:Date;end?:Date) => Date[]

We also need tests to cover the complex functionality of recurrence.

**Comment by timheerwagen** [comment]
During the development of version 1.0.0 I already tried this, but I didn't like it. Because Rrule has a lot of things that we don't need, like its own parsing, etc. In my opinion there must be an easier way to generate this recurring dates.

It is certainly worth creating this functionality from scratch yourself. And besides, I don't like their license.

**Comment by timheerwagen** [comment]
After a lot of hard work, I have now succeeded in providing my own functionality for RRule, including tests. Please have a look at the current development branch to check that I haven't forgotten anything. The feature will be released later in v1.2.0.

**Comment by dbondarchuk** [comment]
That looks great!

**Pull request closed without merging**